phoenix-core/src/main/java/org/apache/phoenix/compile/IndexColumnExpressionCompiler.java

```diff
@@ -0,0 +1,302 @@
+    }
+	
+    /**
+     * Returns the compiled expression
+     * 
+     * @param node
+     *            data table parse node
+     */
+	private Expression getExpression(ParseNode node) throws SQLException {
+		PhoenixConnection connection = this.context.getConnection();
+        PTable indexTable = this.context.getCurrentTable().getTable();
+        NamedTableNode dataTableNode = NamedTableNode.create(null, TableName.create(indexTable.getParentSchemaName().getString(), 
+        		indexTable.getParentTableName().getString()), Collections.<ColumnDef>emptyList());
+        ColumnResolver resolver = FromCompiler.getResolver(dataTableNode, connection);
+		StatementContext context = new StatementContext(this.context.getStatement(), resolver, new Scan(), new SequenceManager(this.context.getStatement()));        
+        ExpressionCompiler expressionCompiler = new ExpressionCompiler(context);
+        return node.accept(expressionCompiler);		
+	}
+
+	/**
+	 * @return true if current table is an index and there is an expression that matches the given node
+	 */
+	private boolean matchesIndexedExpression(ParseNode node) throws SQLException {   
+		if (context.getCurrentTable().getTable().getType()!=PTableType.INDEX) {
+			return false;
+		}
+		DataTableNodeRewriter statementRewriter = new DataTableNodeRewriter();
+        ParseNode dataTableParseNode = node.accept(statementRewriter);
+		Expression dataTableExpression = getExpression(dataTableParseNode);
+		// if the node matches a column expression
+		if (ColumnExpression.class.isAssignableFrom(dataTableExpression.getClass())) {
+			return false;
+		}
+	    for ( PColumn column : this.context.getCurrentTable().getTable().getColumns()) {
+	        if (column.getExpressionStr()!=null) {
+	            ParseNode parseNode = SQLParser.parseCondition(column.getExpressionStr());
+	            Expression expression = getExpression(parseNode);
+	            if (expression.equals(dataTableExpression))
+	                return true;
+	        }
+	    }
+		return false;
+    }
+    
+    private Expression convertAndVisitParseNode(ParseNode node) throws SQLException {
+        DataTableNodeRewriter statementRewriter = new DataTableNodeRewriter();
+        ParseNode dataTableParseNode = node.accept(statementRewriter);
+    	Expression expression = getExpression(dataTableParseNode);
+    	ColumnParseNode columnParseNode = new ColumnParseNode(null, IndexUtil.getIndexColumnName(null,String.valueOf(expression.hashCode())), null);
+    	return visit(columnParseNode);
+    }
+
+	@Override
+    public boolean visitEnter(ComparisonParseNode node) throws SQLException {
+		// do not visit this node if it matches an expression that is indexed,  it will be converted to a ColumnParseNode and processed in visitLeave
+		return matchesIndexedExpression(node) ? false : super.visitEnter(node);
+    }
+
+    @Override
+    public Expression visitLeave(ComparisonParseNode node, List<Expression> children) throws SQLException {
```

Comment:
Having to specialize every visitorLeave is a bit painful. One possibility would be to make the wrapGroupByExpression method protected, override it, and do a super.wrapGroupByExpression() call (and perhaps rename it). I believe that method is called from every visitorLeave method.